src/protocol/step/compact.rs

```diff
@@ -0,0 +1,143 @@
+impl From<&str> for Compact {
+    fn from(id: &str) -> Self {
+        Compact(id.parse().expect("Failed to parse id {id}"))
+    }
+}
+
+impl Display for Compact {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Debug for Compact {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "step={}", self.0)
+    }
+}
+
+const ROOT_STATE: u16 = 0;
+const FALLBACK_STATE: u16 = 65535;
+const FALLBACK_STEP: &str = "fallback";
+const UPGRADE_SEMI_HONEST_STATE: u16 = 65534;
+const UPGRADE_SEMI_HONEST_STEP: &str = "upgrade_semi-honest";
+
+// Hard-coded state map for steps that are narrows but never executed.
+// Such steps are used in many places in the code base, either for convenience of
+// writing compact code, by design, or for tests.
+// For example, `semi_honest::Upgraded` implements `UpgradeContext` trait, but it's
+// `upgrade` method is never executed. For these steps, we need to provide the state,
+// and sub-states that branch off from it, and deliberately ignore them.
+fn static_state_map(state: u16, step: &str) -> u16 {
+    if state == ROOT_STATE && step.starts_with("run-") {
+        // ignore all `run-*` steps
+        // we use `starts_with to avoid regex dependency. should be good enough
+        return ROOT_STATE;
+    } else if step == FALLBACK_STEP {
+        // RBG fallback narrow
+        return FALLBACK_STATE;
+    } else if step == UPGRADE_SEMI_HONEST_STEP || state == UPGRADE_SEMI_HONEST_STATE {
+        // semi-honest's dummy narrow in `UpgradeContext::upgrade()`
+        // ... and any subsequent narrows from this state will be ignored
+        return UPGRADE_SEMI_HONEST_STATE;
+    }
+
+    panic!("cannot narrow with \"{step}\" from state {state}");
+}
+
+// Reverse of `static_state_map` for `Compact::as_ref()`
+fn static_reverse_state_map(state: u16) -> &'static str {
+    if state == ROOT_STATE {
+        return "run-0";
+    } else if state == FALLBACK_STATE {
+        return FALLBACK_STEP;
+    } else if state == UPGRADE_SEMI_HONEST_STATE {
+        return UPGRADE_SEMI_HONEST_STEP;
+    }
+
+    panic!("cannot as_ref() from the invalid state {state}");
+}
+
+//
+// "conditional" steps
+//
+
+impl StepNarrow<crate::protocol::context::semi_honest::UpgradeStep> for Compact {
+    fn narrow(&self, step: &crate::protocol::context::semi_honest::UpgradeStep) -> Self {
+        Self(static_state_map(self.0, step.as_ref()))
+    }
+}
+
+impl StepNarrow<crate::protocol::boolean::random_bits_generator::FallbackStep> for Compact {
+    fn narrow(&self, step: &crate::protocol::boolean::random_bits_generator::FallbackStep) -> Self {
+        Self(static_state_map(self.0, step.as_ref()))
+    }
+}
+
+//
+// steps used in tests
+//
+
+#[cfg(any(feature = "test-fixture", debug_assertions))]
+impl StepNarrow<str> for Compact {
+    fn narrow(&self, step: &str) -> Self {
+        Self(static_state_map(self.0, step))
```

Comment:
static allows mapping to root step, fallback for RGB and upgrade. Which branch is picked for test steps?

Reply:
The first call to `narrow` from TestWorld will have step string "run-0", and the gate state ID is 0, so it's mapped to the root step. Once this is matched, subsequent `narrow` calls will/should not reach `static_state_map` (unless the step is one of dummy or conditional like semi-honest upgrade and fallback). The narrowed step should match one of "real" protocol steps in one of `StepNarrow` implementations.

Obviously, this won't work in unit tests because but that's expected because we only use descriptive gate with unit tests, and we don't generate full steps for unit tests to begin with. So, the root step mapping in `static_state_map` is there specifically for this case.